Improved get_free_u method

ar4s opened the pull request:
Before `get_free_u` might returned negative value (in real life it's not impossible) after this change `get_free_u` work properly. The method respect two side (front, back) and result is a sum of both sides.

mkurek commented on src/ralph/data_center/models/physical.py:
```diff
@@ -190,22 +191,36 @@ def get_orientation_desc(self):
+                    u_list[position - 1] = False
+        fill_u_list(accessories)
+        fill_u_list(dc_assets, lambda obj: obj[1])
+        return Counter(u_list)[True]
```
return `sum(u_list)` will be enough :)

mkurek commented on requirements/test.txt:
```diff
@@ -7,5 +7,6 @@ ddt
 django-plainpasswordhasher
 django_migration_testcase
 factory-boy
+fake-factory
```
where are you using it?

mkurek commented on src/ralph/data_center/models/physical.py:
```diff
@@ -227,6 +242,7 @@ class DataCenterAsset(Asset):
         choices=Orientation(),
         default=Orientation.front.id,
     )
+
```
please remove this line

mkurek commented on src/ralph/data_center/models/physical.py:
```diff
@@ -190,22 +191,36 @@ def get_orientation_desc(self):
+        def fill_u_list(objects, height_of_device=lambda obj: 1):
+            for obj in objects:
+                start = obj[0]
+                end = obj[0] + int(height_of_device(obj))
```
what if someone create device (model) with height of device greater than height of rack (or at least end of device will be greater than end of rack)? IMO this should be `max(self.max_u_height, obj[0] + int(height_of_device(obj)))`

vi4m commented:
http://chris.beams.io/posts/git-commit/

ar4s committed:
Improved get_free_u method

Before ``get_free_u`` might returned negative value (in real life it's not
impossible) after this change ``get_free_u`` work properly. The method
respect two side (front, back) and result is a sum of both sides.

mkurek commented:
:ok:

szok commented:
:ok:

xliiv commented on src/ralph/data_center/tests/test_models.py:
```diff
@@ -272,3 +272,47 @@ def test_change_rack_in_descendants(self):
+        (1, 46, 47, 1),
+        (2, 45, 47, 2),
+        (47, 1, 47, 46),
+        (47, 5, 47, 46),
```
Is this valid?

ar4s replied:
nope :) but is ok - it testing corner case

xliiv replied:
would you add comment for weirdos like me ;P
it could be
`# works ok even when position greater than max_u`

ghost commented:
:ok: (my comment is minor, I'm not blocking this change)
